File: platform-instances/backend-for-frontend/src/main/java/com/sv/youapp/bff/services/impl/DefaultTokenExchangeService.java
package com.sv.youapp.bff.services.impl;

import static com.sv.youapp.bff.utils.RequestUtils.encode256UrlSafe;

import com.sv.youapp.bff.configuration.ServerPort;
import com.sv.youapp.bff.dto.AuthorizationResponse;
import com.sv.youapp.bff.oauth2.spec.AuthorizationCodeRequestSpec.ProofKeyForCodeExchangeRequestSpec;
import com.sv.youapp.bff.oauth2.spec.RequestSpec;
import com.sv.youapp.bff.properties.BackEndForFrontEndProperties;
import com.sv.youapp.bff.services.TokenExchangeService;
import com.sv.youapp.bff.utils.RequestUtils;
import com.sv.youapp.common.authorization.dto.SessionRequest;
import com.sv.youapp.common.authorization.services.SessionStorage;
import java.nio.charset.StandardCharsets;
import java.util.HashSet;
import java.util.Set;
import lombok.NonNull;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.util.UriComponents;
import org.springframework.web.util.UriComponentsBuilder;
import reactor.core.publisher.Mono;

public record DefaultTokenExchangeService(
    @NonNull WebClient webClient,
    @NonNull SessionStorage sessionStorage,
    @NonNull BackEndForFrontEndProperties properties)
    implements TokenExchangeService {

  @Override
  public Mono<AuthorizationResponse> exchange(String code, String state) {
    SessionRequest request = sessionStorage.get(state);
    // IF DOESNT EXITS, CANT CONTINUE
    if (request == null) {
      throw new IllegalStateException("CRSF DETECTED, SESSION EXPIRED OR INVALID");
    }
    return TokenExchangeService.authorizationCode()
        .webClient(webClient)
        .code(code)
        .codeVerifier(request.codeVerifier())
        .basic(properties.clientId(), properties.clientSecret())
        .redirectUri(request.redirectUri())
        .retrieve();
  }

  @Override
  public Mono<Void> start(ServerHttpResponse res) {
    final String state = encode256UrlSafe();
    final String nonce = encode256UrlSafe();
    final String codeVerifier = encode256UrlSafe();
    final String redirectUri = resolveSelf();
    final Set<String> scopes = new HashSet<>();
    UriComponents uri =
        TokenExchangeService.authorizationCodeRequest(properties.url().toString())
            // TODO: RECEIVE FROM APP?
            .clientId(properties().clientId())
            .request((RequestSpec<?> x) -> x.scopes(scopes).state(state).redirectUri(redirectUri))
            .pkce((ProofKeyForCodeExchangeRequestSpec pkce) -> pkce.codeVerifier(codeVerifier))
            .build();
    sessionStorage.save(state, new SessionRequest(state, scopes, redirectUri, codeVerifier, nonce));
    return TokenExchangeService.redirect(res, uri.toUriString());
  }

  private String resolveSelf() {
    String ip = RequestUtils.resolveNonLoopbackAddress();
    UriComponents uri =
        UriComponentsBuilder.fromUriString(String.format("http://%s:%s", ip, ServerPort.getPort()))
            .path("/oauth2/callback")
            .encode(StandardCharsets.UTF_8)
            .build();
    return uri.toString();
  }
}

File: platform-instances/backend-for-frontend/src/main/java/com/sv/youapp/bff/services/impl/DefaultTokenExchangeService.kt
//package com.sv.youapp.bff.services.impl
//
//import com.sv.youapp.bff.services.TokenExchangeService
//import org.springframework.http.HttpHeaders
//import org.springframework.http.MediaType
//import org.springframework.security.config.Customizer
//import org.springframework.web.reactive.function.BodyInserters
//import org.springframework.web.reactive.function.client.WebClient
//import reactor.core.publisher.Mono
//
//class DefaultTokenExchangeService private constructor(val uri: String,
//                                                      val redirectUri: String, val webClient: WebClient): TokenExchangeService {
//
//    companion object {
//        fun builder(): Builder = Builder()
//    }
//
//    class Builder {
//        private var uri: String = "/oauth2/token"
//
//        private var redirectUri: String = "/oauth2/callback"
//
//        private var webClient: TokenWebClientSpec = TokenWebClientSpec()
//
//        fun tokenUri(uri: String) =  apply { this.uri = uri }
//
//        fun redirectUri(url: String) = apply { this.redirectUri = url }
//
//        fun client(customizer: Customizer<TokenWebClientSpec>) = apply { customizer.customize(this.webClient) }
//
//        //fun client(webClient: WebClient) = apply { this.webClient = webClient }
//
//        fun build(): TokenExchangeService {
//            return DefaultTokenExchangeService(uri, redirectUri, webClient.configure())
//        }
//    }
//
//    class TokenWebClientSpec {
//        private var host = "http://authentication-server/"
//
//        fun host(host: String) = apply { this.host = host }
//
//        internal fun configure(): WebClient {
//            return WebClient.builder().baseUrl(host).build()
//        }
//    }
//
//
//
//
//    override fun exchange(code: String): Mono<Map<*, *>?> {
//        return webClient.post().uri(uri)
//            .headers { x: HttpHeaders ->  x.setBasicAuth("oidc-client", "secret") }
//            .contentType(MediaType.APPLICATION_FORM_URLENCODED)
//            .body (
//                BodyInserters.fromFormData("grant_type", "authorization_code")
//                    .with("code", code)
//                    .with("redirect_uri", "http://192.168.1.24:8083/oauth2/callback")
//            ).retrieve()
//            .bodyToMono(Map::class.java)
//    }
//
//    override fun init(): Mono<Void> {
//        TODO("Not yet implemented")
//    }
//
//
//}
//
//

